### ISSUE TITLE
Deployment Strategies

### ISSUE THREAD

**Issue by GandaG** [wrye-bash/wrye-bash#473]
Now that we're halfway to a proper environment setup and building we can stop deploying like cavemen and automate everything!

1. **dropbox** - [*WIP*] - Nightly only
    * Currently we manually copy files to a shared folder. Like cavemen. Disgusting.
    * Automated deployment would necessarily include creating a [dropbox app](https://www.dropbox.com/developers/reference/getting-started#app%20console) with "Full Dropbox" permissions ("App Folder" does not allow folder sharing).
    * The dropbox app provides "client id" and "secret" that should be kept stored in a secure location - we currently only have the discord dev channels as such a place.
    * Dev dropbox credentials would be required to deploy via the dropbox app.
    * The deployment should default to the root of the shared folder, with an optional cli arg to deploy to a branch folder (maybe for testing purposes? Could be useful).

2. **Nexus** - [*WIP*] - Nightly and Production
    * Currently we only upload production files manually. It's a *ton* of work to do this manually, this is probably the highest priority to automate.
    * The [Nexus API](https://app.swaggerhub.com/apis-docs/NexusMods/nexus-mods_public_api_params_in_form_data/1.0#/) does not allow file uploading (yet?) so we must do with a selenium script.
    * It is possible to use cookies to login - we should totally leverage that.
    * There are two major issues with a script for this: nexus uses some overlays that are notoriously hard to rid of (@Infernio care to weigh in?) and nexus keeps changing the upload page html which forces us to keep updating the scripts.
    * Selenium should be run headless to prevent bugs.
    * Selenium's driver for the browser should be downloaded automatically.
    * Production files would be uploaded to the "Main Files" category and nightly to "Updates". Any previous files present should be deleted.

3. **discord** - [*backburner*] - Nightly only
    * Currently @Infernio toils away at updating the post in the *wip-builds* channel with a beautiful changelog for every nightly - is he not entitled to the sweat of his brow? "No!", says @Sharlikran, "it belongs to the users". "No!", says @Arthmoor, "it belongs to the community". "No!", says @FelesNoctis, "it belongs to the devs". I rejected those answers; instead, I chose something different. I chose to automate.
    * If we want to keep a single post on the *wip-builds* channel we'll need a bot to edit it - this requires a place to host the bot I believe, unless it can auto-add itself.
    * If we're going with new posts on that channel we can do something like *LOOT* does, with a webhook, which is substantially easier.
    * The main issue are the changelogs - maybe start keeping a written changelog? Not sure what to do here.

4. **chocolatey** - [*unplanned*] - Nightly and Production
    * No current deployment here.
    * Not sure how feasible this is, but something we should keep in mind.
    * If we're allowed to set package dependencies, we could on the python package and run the source code directly - this would bypass some freezing issues like LOOT problems.

5. **linux repositories** - [*unplanned*] - Nightly and Production
    * No current deployment here.
    * Something for the far away future. Deploying to a *PPA* or to *AUR* would be fun, specially since Bethesda games are doing pretty well in [proton](https://www.protondb.com/search?q=bethesda).

General Considerations:
* All credentials used when deploying should be saved somewhere to avoid devs having to reintroduce them every time. Ideally, the could also be provided by cli args to help with CI integration.
* All deployments should be a one-liner that required minimal to no interaction from the dev - some of these could be lengthy and forcing the dev to watch a log is no fun.
* Since these deployments are all meant to be used every time (with exceptions for nightly/production-only) we could use an umbrella script to facilitate deployments.

Related Branches: [ganda-260-deploy](https://github.com/wrye-bash/wrye-bash/compare/ganda-260-deploy)

I think you all know my [thoughts on this](https://github.com/wrye-bash/wrye-bash/commit/67eab346ec4d2f9ab4e9414768a6dd266ccfbd12#commitcomment-35275004).

Discuss!

**Comment by Utumno** [comment]
> Currently we only upload production files manually. It's a ton of work to do this manually, this is probably the highest priority to automate.

I would agree let's have 307 out that way - assigning the rest to 308 although they may be done sooner - priority remains wx3/phoenix/py3 update for 307

**Comment by GandaG** [comment]
> let's have 307 out that way

307 Beta 4 was uploaded successfully via the scripts in *ganda-260-deploy*. There was a small hitch that is being worked on due to Infernio's Chrome freaking out a bit :P

**Comment by GandaG** [comment]
Small update, deployment branch has been split per target

**Comment by Infernio** [comment]
*(Dropped C-enhancement since this is primarily a discussion issue for now - readd enhancement in 308 once discussion is decided)*

**Comment by Infernio** [comment]
For tracking purposes, the upstream issue that's blocking us from just uploading our nice, script-built 7z files is this: https://github.com/actions/upload-artifact/issues/39

**Comment by Infernio** [comment]
Ref https://feedback.nexusmods.com/posts/11/mod-upload-api
